src/bom/chainsop.rs

```diff
@@ -0,0 +1,872 @@
+///
+///     In addition, chainsop provides automatic creation and management of
+///     temporary files used in the above.
+///
+///     The chainsop package provides more direct support for incrementally
+///     building the set of commands with outputs; the subprocess crate would
+///     require more discrete management and building of a Vec<Exec>.
+///
+///     The chainsop allows elements of the chain to be local functions called in
+///     the proper sequence of operations and for elements of the chain to be
+///     disabled prior to actual execution (where they are skipped).
+///
+///     The subprocess crate provides more features for handling stdout/stderr
+///     redirection, non-blocking and timed sub-process waiting, and interaction
+///     with the sub-process.
+///
+///     Summary: significant overlap in capabilities with slightly different
+///     use-case targets and features.
+///
+/// * duct (https://github.com/oconner663/duct.rs
+///
+///     Lightweight version of the subprocess crate
+///
+/// * cargo-make, devrc, rhiz, run-cli, naumann, yamis
+///
+///    Task runners, requiring an external specification of the commands and no
+///    support for chaining inputs/outputs.  These could be written on top of
+///    chainsop.
+///
+/// * steward crate (https://crates.io/crates/steward)
+///
+///    Useful for running multiple commands and allows dependency management, but
+///    not input/output chaining or incremental command building.  Does support
+///    other features like environment control and process pools.  Closer to
+///    chainsop than the task runners, but again, this could be written on top of
+///    chainsop.
+
+use std::cell::{RefCell, RefMut};
+use std::env::current_dir;
+use std::ffi::{OsString};
+use std::fmt;
+use std::path::{Path,PathBuf};
+use std::process;
+use std::rc::Rc;
+
+
+/// Designates a type of file that can be identified by name on the command line.
+#[non_exhaustive]
+#[derive(Clone, Debug)]
+pub enum NamedFile {
+    /// Create a temporary file; str is suffix to give temporary filename
+    Temp(String),
+
+    /// Actual filename (may or may not currently exist)
+    Actual(PathBuf),
+
+    // Multiple input files: not yet supported.  It is ostensibly better to
+    // represent them here because all input files should share the same FileSpec
+    // enum type, but multiple output files isn't really supported for chaining
+    // to the next input...
+    //
+    // Actuals(Vec<PathBuf>),
+
+    /// glob search in specified dir for all matching files
+    GlobIn(PathBuf, String),
+
+    /// allowed on initial construction, but an error for execute
+    TBD
+}
+
+impl NamedFile {
+    /// Generates the designation indicating the need for a temporary file with
+    /// the specified suffix.  If no particular suffix is needed, a blank suffix
+    /// value should be specified.
+    pub fn temp<T>(suffix: T) -> NamedFile
+    where String: From<T>
```

Comment:
According to https://doc.rust-lang.org/nightly/std/convert/trait.From.html, "One should always prefer implementing  `From` over `Into` ..."

Reply:
That is implementing - not using. I think they provide that guidance because you get the other one for free. That doesn't say anything about not using Into

Reply:
Good point: implement strongly, require weakly.  I'll change all these occurrences.